web-ng/src/app/components/share-dialog/share-dialog.component.ts

```diff
@@ -56,52 +77,107 @@ export class ShareDialogComponent {
     this.subscription.add(
       this.projectService
         .getActiveProject$()
+        .pipe(take(1))
         .subscribe(p => this.onProjectLoaded(p))
     );
   }
 
-  /**
-   * Update ACL and projectId when project is loaded.
-   */
-  private onProjectLoaded(project: Project): void {
-    this.projectId = project.id;
-    this.acl.next(
-      project.acl
-        .entrySeq()
-        .toList()
-        .sortBy(pair => pair[0])
-    );
-  }
-
   /**
    * Add/update user role when email address is entered and add is clicked or
    * enter is pressed.
    */
   onAddUserSubmit(): void {
-    if (!this.projectId) {
+    // UI is hidden until project is loaded, so this should never happen.
+    if (!this.projectId || !this.acl) {
       return;
     }
-    // TODO: Allow setting role.
-    this.projectService.updateRole(
-      this.projectId,
-      this.addUserForm.value['email'],
-      this.addUserForm.value['role']
-    );
-    // TODO: Show saving / saved status.
+    const { email, role } = this.addUserForm.value;
+
+    // Add new email/role and update change state. Validation rules prevent
```

Comment:
just wondering: i find it so crazy that we have only validation on client side.. firebase does not do any checks, does it? Do we care?

Reply:
Yes, it is, and yes we do! Firebase is schemaless, so there are no constraints on the structure/contents of data. In the future, we may implement our own microservice and respective API that will only allow valid data to be submitted.

Reply:
...but for now in demo-land it's something we have to live with.

Reply:
nice^^